### ISSUE TITLE
CH552, sdcc and delay.c

### ISSUE THREAD

**Issue by dfxhub** [wagiminator/AVR-Programmer#4]
Hi! Delays are not correct.

sdcc 4.2.0 compiles your implemetation to this:

```
;	src/delay.c:25: SAFE_MOD++;                     // 2 Fsys cycles, for higher Fsys, add operation here
	mov	a,_SAFE_MOD
	inc	a
	mov	_SAFE_MOD,a
```

and it is not 2 Fsys cycles, so f.e. DLY_ms(1000) at 24MHz becomes ~2+ secs. While this is more correct (about 1.200 secs at 24MHz) (as of [Blinkinlabs](https://github.com/Blinkinlabs) or [official SDK](https://github.com/HonghongLu/CH554/blob/2e219d837e08a15974601d5d01eccc0c11b35b12/official%20ch554%20evt%20pcb/EXAM/Public/Debug.C#L65C8-L65C8)):

```
;	src/delay.c:25: ++SAFE_MOD;                     // 2 Fsys cycles, for higher Fsys, add operation here
	inc	_SAFE_MOD
```
Thanks.

**Comment by wagiminator** [comment]
Yes, I know, this delay routine is very inaccurate. It also varies because with the 8051 the number of clock cycles for jumps can only be predicted exactly under certain conditions. Depending on the version, the compiler then does the rest with its questionable results. For precise delays you have to use a timer or the touchkey interrupt flag.
In this case, however, the delay routine should not and does not have to be precise.

**Comment by dfxhub** [comment]
I agree in that for applications with strong timings requirements delays based on timers or any other hardware are preferred. But why not to use right syntax for particular compiler that leads to more reasonable results? Of cause, IMO, more right way for software delays with predictable results is using assembler and not to trust to compiler, or at least to check for it.

Thanks again for your helpful projects.

**Comment by wagiminator** [comment]
I've made a few changes so that the delays are hopefully more precise now.
